Resolve CSS gradient color stop positions to Lengths at style builder time

weinig opened the pull request:
#### 0bc9f1178d891a5a0987511a69102304e60cdd7b
<pre>
Resolve CSS gradient color stop positions to Lengths at style builder time
<a href="https://bugs.webkit.org/show_bug.cgi?id=275399">https://bugs.webkit.org/show_bug.cgi?id=275399</a>

Reviewed by Antti Koivisto.

Resolve gradient stop positions to Lengths at style builder time.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-linear-radial-conic-gradient-001-expected.txt:
    - Update results for additional passing cases.

* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
    - Update to resolve stop positions when style building. Now that this is
      done here, we need to update caching logic for cases that used to be
      in StyleGradientImage.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
    - Expose zoomAdjustedPixelValueForLength for use by StyleGradientImage.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
    - Use properly typed stop positions to resolve to computed styles
      and rendering.

Canonical link: <a href="https://commits.webkit.org/280120@main">https://commits.webkit.org/280120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac50a7c5742fe56c88c2b82a9280d8eb16c3f245

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55631 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34954 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6061 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57757 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6260 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44797 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4162 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57660 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32878 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47948 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25929 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29665 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5277 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4205 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60206 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30785 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5670 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52228 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31870 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51716 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12364 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32951 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31617 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSGradientValue.cpp:
```diff
@@ -67,9 +112,12 @@ RefPtr<StyleImage> CSSLinearGradientValue::createStyleImage(Style::BuilderState&
 
     auto styleImage = StyleGradientImage::create(
         // FIXME: The parameters to LinearData should convert down to a non-CSS specific type here (e.g. Length, double, etc.).
```
aren't you fixing these FIXMEs?

weinig replied:
That's the next PR. Doing both at once turned into too huge a change.

anttijk commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -43,24 +43,21 @@
-        // FIXME: Do we need handle calc() here?
-        if (stop.position && stop.position->isFontRelativeLength())
-            return false;
         if (stop.color && stop.color->containsCurrentColor())
             return false;
```
do you have plans for currentcolor?

weinig replied:
I believe currentColor works (I think you fixed it!). This is just about caching.

anttijk replied:
Oh nice. I somehow thought gradients were still broken. I suppose this cache is about paint time caching and in could be allowed by including color property to the cache key.

Commit:
Resolve CSS gradient color stop positions to Lengths at style builder time
https://bugs.webkit.org/show_bug.cgi?id=275399

Reviewed by Antti Koivisto.

Resolve gradient stop positions to Lengths at style builder time.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-linear-radial-conic-gradient-001-expected.txt:
    - Update results for additional passing cases.

* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
    - Update to resolve stop positions when style building. Now that this is
      done here, we need to update caching logic for cases that used to be
      in StyleGradientImage.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
    - Expose zoomAdjustedPixelValueForLength for use by StyleGradientImage.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
    - Use properly typed stop positions to resolve to computed styles
      and rendering.

Canonical link: https://commits.webkit.org/280120@main
